### PR TITLE
feat: impl `system.cnch_detached_parts` table

### PR DESCRIPTION
**Pull request by skyoct**
### Changelog category :
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
 impl cnch detach table
 fixes: #1670

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by skyoct** [comment]
@smmsmm1988 Hi, Could you please help review this?

**Comment by smmsmm1988** [comment]
> @smmsmm1988 Hi, Could you please help review this?

@skyoct LGTM. I suggest you add some test cases for this new system table.

**Comment by smmsmm1988** [comment]
@zeromem please help double review this pr

File: src/Storages/System/StorageSystemCnchDetachedParts.cpp [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        {"min_block_number", std::make_shared<DataTypeNullable>(std::make_shared<DataTypeInt64>())},
+        {"max_block_number", std::make_shared<DataTypeNullable>(std::make_shared<DataTypeInt64>())},
+        {"level", std::make_shared<DataTypeNullable>(std::make_shared<DataTypeUInt32>())}
+    };
+}
+
+void StorageSystemCnchDetachedParts::fillData(MutableColumns & res_columns, ContextPtr context, const SelectQueryInfo & query_info) const
+{
+
+   auto query_context = Context::createCopy(context);
```
Comment (smmsmm1988) [inline src/Storages/System/StorageSystemCnchDetachedParts.cpp:32]: why not use context directly but creating a copy?

Comment (skyoct) [reply src/Storages/System/StorageSystemCnchDetachedParts.cpp:32]: @smmsmm1988 In the constructor of CnchAttachProcessor, the type of context is ContextMutablePtr, and it seems that there is no direct way to pass the current context as a parameter.  Is there a more elegant way to write this?

Comment (smmsmm1988) [reply src/Storages/System/StorageSystemCnchDetachedParts.cpp:32]: OK, then it is necessary. Please fix the indent typo of this line.

Comment (skyoct) [reply src/Storages/System/StorageSystemCnchDetachedParts.cpp]: ok

**Comment by skyoct** [comment]
@smmsmm1988 @zeromem Hi 
image
 It seems there is an issue with the CI node.

**Comment by smmsmm1988** [comment]
> @smmsmm1988 @zeromem image Hi, It seems there is an issue with the CI node.

@skyoct you can rerun the test

File: src/Storages/System/StorageSystemCnchDetachedParts.cpp [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+{
+
+    auto query_context = Context::createCopy(context);
+
+    auto cnch_catalog = context->getCnchCatalog();
+
+    ASTPtr where_expression = query_info.query->as<ASTSelectQuery>()->where();
+    std::map<String, String> column_to_value;
+
+    collectWhereClausePredicate(where_expression, column_to_value);
```
Comment (zeromem) [inline src/Storages/System/StorageSystemCnchDetachedParts.cpp:39]: 参考 StorageSystemCnchParts 这里一般调用的是collectWhereORClausePredicate

File: tests/queries/4_cnch_stateless/65007_cnch_detached_parts.sql [hunk 3]
```diff
@@ -0,0 +1,11 @@
+DROP DATABASE IF EXISTS test_db;
```
Comment (zeromem) [inline tests/queries/4_cnch_stateless/65007_cnch_detached_parts.sql:1]: CI testcase 一般不指定 database。创建表之后，如果需要查询用到 database，可以用 `currentDatabase()`

File: src/Storages/System/StorageSystemCnchDetachedParts.cpp [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+            continue;
+
+        if (auto * cnch_merge_tree = dynamic_cast<StorageCnchMergeTree *>(storage.get()))
+        {
+            PartitionCommand cmd;
+            CnchAttachProcessor processor(*cnch_merge_tree, cmd, query_context);
+            auto detached_parts = processor.getDetachedParts(AttachFilter::createPartsFilter());
+            for (const auto & parts : detached_parts)
+            {
+                for (const auto & part : parts)
```
Comment (zeromem) [inline src/Storages/System/StorageSystemCnchDetachedParts.cpp:102]: 因为这里是 collect 好的 part chain，这里最好给表里面加个字段 part_type 标识一下。可以参考 system.cnch_parts 的 part_type。可以复用 visible 和 invisible 的含义。

Comment (zeromem) [reply src/Storages/System/StorageSystemCnchDetachedParts.cpp:102]: btw，CI 里面最好也加上 partial part 的 test。一个 part `p00` 做完 mutation 之后，会产生一个 partial part 进而形成 part chain。 `p00 <- p01`。 这时候  `p01`是 visible part，`p00` 是 invisible part。查询 system.cnch_detached_parts 需要区分出来这种case，count() 预期为 2，count() where part_type = 'VisiblePart' 预期为 1

**Comment by skyoct** [comment]
@zeromem 你好，能帮忙重新review一下吗

**Review by zeromem — APPROVED** [review]

**Pull request merged**